scipio/src/sys/rusage.rs

```diff
@@ -0,0 +1,123 @@
+    pub ru_oublock: u64,
+    /// The number of times a context switch resulted due to a process
+    /// voluntarily giving up the processor before its time slice
+    /// was completed (usually to await availability of a
+    /// resource).
+    pub ru_nvcsw: u64,
+    /// The number of times a context switch resulted due to a higher
+    /// priority process becoming runnable or because the current
+    /// process exceeded its time slice.
+    pub ru_nivcsw: u64,
+}
+
+impl ResourceUsage {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn get() -> Self {
+        Self::from(getrusage())
+    }
+
+    pub fn runtime() -> Self {
+        let r = getrusage();
+        let mut u = ResourceUsage::new();
+
+        u.ru_utime = from_timeval(r.ru_utime).as_nanos() as u64;
+        u.ru_stime = from_timeval(r.ru_stime).as_nanos() as u64;
+        u.ru_nvcsw = r.ru_nvcsw as u64;
+
+        return u;
+    }
+}
+
+impl Default for ResourceUsage {
+    fn default() -> Self {
+        ResourceUsage {
+            ru_utime: 0,
+            ru_stime: 0,
+            ru_maxrss: 0,
+            ru_minflt: 0,
+            ru_majflt: 0,
+            ru_inblock: 0,
+            ru_oublock: 0,
+            ru_nvcsw: 0,
+            ru_nivcsw: 0,
+        }
+    }
+}
+
+impl From<libc::rusage> for ResourceUsage {
+    fn from(r: libc::rusage) -> Self {
+        ResourceUsage {
+            ru_utime: from_timeval(r.ru_utime).as_nanos() as u64,
+            ru_stime: from_timeval(r.ru_stime).as_nanos() as u64,
+            ru_maxrss: r.ru_maxrss as u64,
+            ru_minflt: r.ru_maxrss as u64,
+            ru_majflt: r.ru_majflt as u64,
+            ru_inblock: r.ru_inblock as u64,
+            ru_oublock: r.ru_oublock as u64,
+            ru_nvcsw: r.ru_nvcsw as u64,
+            ru_nivcsw: r.ru_nivcsw as u64,
+        }
+    }
+}
+
+impl Sub for ResourceUsage {
+    type Output = ResourceUsage;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        ResourceUsage {
+            ru_utime: self.ru_utime - rhs.ru_utime,
+            ru_stime: self.ru_stime - rhs.ru_stime,
+            ru_maxrss: self.ru_maxrss - rhs.ru_maxrss,
+            ru_minflt: self.ru_minflt - rhs.ru_minflt,
+            ru_majflt: self.ru_majflt - rhs.ru_majflt,
+            ru_inblock: self.ru_inblock - rhs.ru_inblock,
+            ru_oublock: self.ru_oublock - rhs.ru_oublock,
+            ru_nvcsw: self.ru_nvcsw - rhs.ru_nvcsw,
+            ru_nivcsw: self.ru_nivcsw - rhs.ru_nivcsw,
+        }
+    }
+}
+
+/// As far as impl From<libc::timeval> for Duration is not allowed.
+fn from_timeval(v: libc::timeval) -> Duration {
+    Duration::from_micros((v.tv_sec * 1_000_000 + v.tv_usec) as u64)
+}
+
+fn getrusage() -> libc::rusage {
+    use core::mem::MaybeUninit;
+    let mut s0 = MaybeUninit::<libc::rusage>::uninit();
+    let err = unsafe { libc::getrusage(libc::RUSAGE_THREAD, s0.as_mut_ptr()) };
+    if err != 0 {
```

Comment:
assert_eq is better here. 

Btw, did you run `cargo clippy` ?

Reply:
Wiped